embulk-input-s3/src/main/java/org/embulk/input/s3/AlwaysRetryable.java

```diff
@@ -64,6 +77,17 @@ public T call() throws Exception
     @Override
     public boolean isRetryableException(Exception exception)
     {
+        // Always retry on client exceptions caused by IOException
+        if (exception.getCause() instanceof IOException) {
+            return true;
+        }
+        // No retry on a subset of service exceptions
+        if (exception instanceof AmazonServiceException) {
+            AmazonServiceException ase = (AmazonServiceException) exception;
+            if (isNonRetryableServiceException(ase)) {
```

Comment:
`isRetryable()` call `isNonRetryable...` is too much hairy, plus the implementation of `isNonRetryable...` is only a boolean expression, it can be placed inline on `isRetryable` with a complement.